File: ts/tests/webex-http.test.ts
import {
  sendMessage,
  sendCard,
  getMessage,
  getAttachmentAction,
  ADAPTIVE_CARD_CONTENT_TYPE,
} from '../src/webex';

function mockFetch(status: number, body: Record<string, unknown>): typeof fetch {
  return (async (_url: string | URL | Request, _init?: RequestInit) => ({
    ok: status >= 200 && status < 300,
    status,
    text: async () => JSON.stringify(body),
    json: async () => body,
  })) as unknown as typeof fetch;
}

function capturingFetch(
  status: number,
  body: Record<string, unknown>,
): { fetchFn: typeof fetch; captured: { url: string; init: RequestInit }[] } {
  const captured: { url: string; init: RequestInit }[] = [];
  const fetchFn = (async (url: string | URL | Request, init?: RequestInit) => {
    captured.push({ url: url as string, init: init! });
    return {
      ok: status >= 200 && status < 300,
      status,
      text: async () => JSON.stringify(body),
      json: async () => body,
    };
  }) as unknown as typeof fetch;
  return { fetchFn, captured };
}

describe('sendMessage HTTP', () => {
  it('sends correct payload and headers', async () => {
    const { fetchFn, captured } = capturingFetch(200, { id: 'msg-1' });
    const result = await sendMessage('tok123', 'user@example.com', 'hello', fetchFn);

    expect(result).toEqual({ id: 'msg-1' });
    expect(captured.length).toBe(1);
    const body = JSON.parse(captured[0].init.body as string);
    expect(body.toPersonEmail).toBe('user@example.com');
    expect(body.text).toBe('hello');
    expect(captured[0].init.headers).toEqual(
      expect.objectContaining({ Authorization: 'Bearer tok123' }),
    );
  });

  it('throws on HTTP error', async () => {
    const fetchFn = mockFetch(401, { message: 'unauthorized' });
    await expect(sendMessage('badtoken', 'user@example.com', 'hello', fetchFn)).rejects.toThrow(
      '401',
    );
  });
});

describe('sendCard HTTP', () => {
  it('sends card attachment', async () => {
    const card = { type: 'AdaptiveCard', body: [{ type: 'TextBlock', text: 'Hello' }] };
    const { fetchFn, captured } = capturingFetch(200, { id: 'card-1' });
    const result = await sendCard('tok', 'user@x.com', 'fallback', card, fetchFn);

    expect(result).toEqual({ id: 'card-1' });
    const body = JSON.parse(captured[0].init.body as string);
    expect(body.text).toBe('fallback');
    expect(body.toPersonEmail).toBe('user@x.com');
    expect(body.attachments).toHaveLength(1);
    expect(body.attachments[0].contentType).toBe(ADAPTIVE_CARD_CONTENT_TYPE);
    expect(body.attachments[0].content).toEqual(card);
  });
});

describe('getMessage HTTP', () => {
  it('fetches message by ID', async () => {
    const msgData = { id: 'msg-1', text: 'hello', attachments: [] };
    const fetchFn = mockFetch(200, msgData);
    const result = await getMessage('tok', 'msg-1', fetchFn);
    expect(result).toEqual(msgData);
  });

  it('throws on not found', async () => {
    const fetchFn = mockFetch(404, {});
    await expect(getMessage('tok', 'bad-id', fetchFn)).rejects.toThrow('404');
  });
});

describe('getAttachmentAction HTTP', () => {
  it('fetches action by ID', async () => {
    const actionData = { id: 'act-1', type: 'submit', inputs: { name: 'test' } };
    const fetchFn = mockFetch(200, actionData);
    const result = await getAttachmentAction('tok', 'act-1', fetchFn);
    expect(result).toEqual(actionData);
  });
});

File: rust/tests/webex_http_test.rs
use serde_json::json;
use std::sync::Mutex;
use wiremock::matchers::{header, method, path};
use wiremock::{Mock, MockServer, ResponseTemplate};
use wgrok::{_set_attachment_actions_url, _set_messages_url};

static URL_LOCK: Mutex<()> = Mutex::new(());

#[tokio::test]
async fn test_webex_http_all() {
    let _lock = URL_LOCK.lock().unwrap();

    // -- send_message --
    {
        let server = MockServer::start().await;
        _set_messages_url(Some(format!("{}/messages", server.uri())));

        Mock::given(method("POST"))
            .and(path("/messages"))
            .and(header("Authorization", "Bearer tok123"))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({"id": "msg-1"})))
            .mount(&server)
            .await;

        let client = reqwest::Client::new();
        let result = wgrok::send_message("tok123", "user@example.com", "hello", &client)
            .await
            .unwrap();
        assert_eq!(result["id"], "msg-1");

        _set_messages_url(None);
    }

    // -- send_message error --
    {
        let server = MockServer::start().await;
        _set_messages_url(Some(format!("{}/messages", server.uri())));

        Mock::given(method("POST"))
            .and(path("/messages"))
            .respond_with(ResponseTemplate::new(401).set_body_string("unauthorized"))
            .mount(&server)
            .await;

        let client = reqwest::Client::new();
        let result = wgrok::send_message("badtoken", "user@example.com", "hello", &client).await;
        assert!(result.is_err());
        assert!(result.unwrap_err().contains("401"));

        _set_messages_url(None);
    }

    // -- send_card --
    {
        let server = MockServer::start().await;
        _set_messages_url(Some(format!("{}/messages", server.uri())));

        Mock::given(method("POST"))
            .and(path("/messages"))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({"id": "card-1"})))
            .mount(&server)
            .await;

        let card = json!({"type": "AdaptiveCard", "body": []});
        let client = reqwest::Client::new();
        let result = wgrok::send_card("tok", "user@x.com", "fallback", &card, &client)
            .await
            .unwrap();
        assert_eq!(result["id"], "card-1");

        _set_messages_url(None);
    }

    // -- get_message --
    {
        let server = MockServer::start().await;
        _set_messages_url(Some(format!("{}/messages", server.uri())));

        Mock::given(method("GET"))
            .and(path("/messages/msg-1"))
            .respond_with(
                ResponseTemplate::new(200).set_body_json(json!({"id": "msg-1", "text": "hello"})),
            )
            .mount(&server)
            .await;

        let client = reqwest::Client::new();
        let result = wgrok::get_message("tok", "msg-1", &client).await.unwrap();
        assert_eq!(result["id"], "msg-1");
        assert_eq!(result["text"], "hello");

        _set_messages_url(None);
    }

    // -- get_attachment_action --
    {
        let server = MockServer::start().await;
        _set_attachment_actions_url(Some(format!(
            "{}/attachment/actions",
            server.uri()
        )));

        Mock::given(method("GET"))
            .and(path("/attachment/actions/act-1"))
            .respond_with(
                ResponseTemplate::new(200).set_body_json(
                    json!({"id": "act-1", "type": "submit", "inputs": {"name": "test"}}),
                ),
            )
            .mount(&server)
            .await;

        let client = reqwest::Client::new();
        let result = wgrok::get_attachment_action("tok", "act-1", &client)
            .await
            .unwrap();
        assert_eq!(result["id"], "act-1");
        assert_eq!(result["type"], "submit");

        _set_attachment_actions_url(None);
    }
}

#[tokio::test]
async fn test_retry_after_retries_on_429_then_succeeds() {
    let _lock = URL_LOCK.lock().unwrap();

    let server = MockServer::start().await;
    _set_messages_url(Some(format!("{}/messages", server.uri())));

    // First attempt: 429 with Retry-After
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(
            ResponseTemplate::new(429)
                .append_header("Retry-After", "1")
                .set_body_string("rate limited"),
        )
        .up_to_n_times(1)
        .mount(&server)
        .await;

    // Second attempt: 200 OK
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({"id": "msg-retry-1"})))
        .mount(&server)
        .await;

    let client = reqwest::Client::new();
    let result = wgrok::send_message("tok123", "user@example.com", "hello", &client)
        .await
        .unwrap();
    assert_eq!(result["id"], "msg-retry-1");

    _set_messages_url(None);
}

#[tokio::test]
async fn test_retry_after_retries_multiple_429s() {
    let _lock = URL_LOCK.lock().unwrap();

    let server = MockServer::start().await;
    _set_messages_url(Some(format!("{}/messages", server.uri())));

    // First attempt: 429 with Retry-After: 1
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(
            ResponseTemplate::new(429)
                .append_header("Retry-After", "1")
                .set_body_string("rate limited 1"),
        )
        .up_to_n_times(1)
        .mount(&server)
        .await;

    // Second attempt: 429 with Retry-After: 2
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(
            ResponseTemplate::new(429)
                .append_header("Retry-After", "2")
                .set_body_string("rate limited 2"),
        )
        .up_to_n_times(1)
        .mount(&server)
        .await;

    // Third attempt: 200 OK
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({"id": "msg-retry-2"})))
        .mount(&server)
        .await;

    let client = reqwest::Client::new();
    let result = wgrok::send_message("tok123", "user@example.com", "hello", &client)
        .await
        .unwrap();
    assert_eq!(result["id"], "msg-retry-2");

    _set_messages_url(None);
}

#[tokio::test]
async fn test_retry_after_raises_after_max_retries() {
    let _lock = URL_LOCK.lock().unwrap();

    let server = MockServer::start().await;
    _set_messages_url(Some(format!("{}/messages", server.uri())));

    // All attempts: 429 (4 total attempts before giving up)
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(
            ResponseTemplate::new(429)
                .append_header("Retry-After", "1")
                .set_body_string("rate limited"),
        )
        .mount(&server)
        .await;

    let client = reqwest::Client::new();
    let result = wgrok::send_message("tok123", "user@example.com", "hello", &client).await;
    assert!(result.is_err());
    assert!(result.unwrap_err().contains("429"));

    _set_messages_url(None);
}

#[tokio::test]
async fn test_retry_after_uses_retry_after_header_value() {
    let _lock = URL_LOCK.lock().unwrap();

    let server = MockServer::start().await;
    _set_messages_url(Some(format!("{}/messages", server.uri())));

    // First attempt: 429 with Retry-After: 1 (we'll sleep for this)
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(
            ResponseTemplate::new(429)
                .append_header("Retry-After", "1")
                .set_body_string("rate limited"),
        )
        .up_to_n_times(1)
        .mount(&server)
        .await;

    // Second attempt: 200 OK
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({"id": "msg-retry-3"})))
        .mount(&server)
        .await;

    let client = reqwest::Client::new();
    let result = wgrok::send_message("tok123", "user@example.com", "hello", &client)
        .await
        .unwrap();
    assert_eq!(result["id"], "msg-retry-3");

    _set_messages_url(None);
}

#[tokio::test]
async fn test_retry_after_defaults_retry_after_to_1_when_missing() {
    let _lock = URL_LOCK.lock().unwrap();

    let server = MockServer::start().await;
    _set_messages_url(Some(format!("{}/messages", server.uri())));

    // First attempt: 429 without Retry-After header (should default to 1)
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(ResponseTemplate::new(429).set_body_string("rate limited"))
        .up_to_n_times(1)
        .mount(&server)
        .await;

    // Second attempt: 200 OK
    Mock::given(method("POST"))
        .and(path("/messages"))
        .respond_with(ResponseTemplate::new(200).set_body_json(json!({"id": "msg-retry-4"})))
        .mount(&server)
        .await;

    let client = reqwest::Client::new();
    let result = wgrok::send_message("tok123", "user@example.com", "hello", &client)
        .await
        .unwrap();
    assert_eq!(result["id"], "msg-retry-4");

    _set_messages_url(None);
}

#[tokio::test]
async fn test_retry_after_get_json() {
    let _lock = URL_LOCK.lock().unwrap();

    let server = MockServer::start().await;
    _set_messages_url(Some(format!("{}/messages", server.uri())));

    // First attempt: 429 with Retry-After
    Mock::given(method("GET"))
        .and(path("/messages/msg-retry-5"))
        .respond_with(
            ResponseTemplate::new(429)
                .append_header("Retry-After", "1")
                .set_body_string("rate limited"),
        )
        .up_to_n_times(1)
        .mount(&server)
        .await;

    // Second attempt: 200 OK
    Mock::given(method("GET"))
        .and(path("/messages/msg-retry-5"))
        .respond_with(
            ResponseTemplate::new(200)
                .set_body_json(json!({"id": "msg-retry-5", "text": "retrieved"})),
        )
        .mount(&server)
        .await;

    let client = reqwest::Client::new();
    let result = wgrok::get_message("tok", "msg-retry-5", &client)
        .await
        .unwrap();
    assert_eq!(result["id"], "msg-retry-5");
    assert_eq!(result["text"], "retrieved");

    _set_messages_url(None);
}
